src/compiler/checker.ts

```diff
@@ -14058,8 +14058,16 @@ namespace ts {
             if (!(links.flags & NodeCheckFlags.TypeChecked)) {
                 // Check whether the file has declared it is the default lib,
                 // and whether the user has specifically chosen to avoid checking it.
-                if (node.isDefaultLib && compilerOptions.skipDefaultLibCheck) {
-                    return;
+                if (compilerOptions.skipDefaultLibCheck) {
+                    if (node.isDefaultLib) {
+                        return;
+                    }
+
+                    // If the user specified '--noLib' and a file has a '/// <reference no-default-lib="true"/>',
+                    // then we should treat that file as a default lib.
+                    if (compilerOptions.noLib && node.hasNoDefaultLib) {

```

Comment:
i do not think you need "compilerOptions.noLib" check. having a file with the NoDefaultLib comment is equivalent to --noLib.

I propose we simplify this whole thing, and remove sourceFile.isDefaultLib, as it is not clear why we have it. we should have one concept, is this file a default lib, and that is determined by the flag, the lib.d.ts we add has the flag anyways, so things should just work without the additional flag we are passing through.

we can change our test, to not include the /// <no-default-lib> comment. and just pass `--noLib` instead, this way it still type checks, but does not trigger this condition.